src/mlpack/methods/ann/loss_functions/mean_absolute_percentage_error_impl.hpp

```diff
@@ -0,0 +1,70 @@
+/**
+ * @file methods/ann/loss_functions/mean_absolute_percentage_error_impl.hpp
+ * @author Aakash Kaushik
+ *
+ * Implementation of the Mean Absolute Percentage Error function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LOSS_FUNCTION_MEAN_ABSOLUTE_PERCENTAGE_ERROR_IMPL_HPP
+#define MLPACK_METHODS_ANN_LOSS_FUNCTION_MEAN_ABSOLUTE_PERCENTAGE_ERROR_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "mean_absolute_percentage_error.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+MeanAbsolutePercentageError<InputDataType, OutputDataType>::MeanAbsolutePercentageError(
+  const bool reduction) : reduction(reduction)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType>
+typename InputType::elem_type
+MeanAbsolutePercentageError<InputDataType, OutputDataType>::Forward(
+    const InputType& input,
+    const TargetType& target)
+{
+  InputType loss = arma::abs((input - target) / target);
+  typename InputType::elem_type lossSum = arma::accu(loss);
+
+  if (reduction)
+    return lossSum;
+    
+  return lossSum / target.n_cols;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType, typename OutputType>
+void MeanAbsolutePercentageError<InputDataType, OutputDataType>::Backward(
+    const InputType& input,
+    const TargetType& target,
+    OutputType& output)
+
+{
+  output = 1 / target;
```

Comment:
Maybe I'm wrong but I think the derivative is:

`- 1 / target` if input < target
`1 / target` otherwise

Reply:
I think that is not true as we don't actually have a numerical limit on target:
Suppose target is a vector or even a scalar then input<target doesn't really mean that input is smaller than zero, but if that was the case and the input was smaller than zero then MAPE  would become `( - input - target ) / target` and then the derivate could be `- 1 / target` or otherwise as you specified.

Reply:
Hi, going by this [answer](https://stats.stackexchange.com/questions/312737/mean-absolute-error-mae-derivative), I think you are right, can you check this out ?

Reply:
I have edited and pushed the changes for this one too.

Reply:
please ignore the last commit i am actually a bit confused on how to implement this  `- 1 / target `for input < target and this`1 / target` otherwise.
1. how should i compare these two matrix ?
2. how should i assign the output ?

Reply:
Hi @zoq,
I have made all the changes that were pointed out, Can you give this a final review ?